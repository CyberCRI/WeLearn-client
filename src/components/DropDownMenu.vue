<script setup lang="ts">
import { onMounted, onUnmounted } from 'vue';
import ChevronDownVue from '@/components/icons/ChevronDown.vue';

import { ref } from 'vue';

defineProps<{ title: string; inElement?: boolean; closeOnClick?: boolean }>();
const isVisible = ref(false);
const dropdownElement = ref<HTMLElement>();

const handleClickOutside = ({ target }: MouseEvent) => {
  const targetElement = dropdownElement.value;

  if (targetElement && !targetElement.contains(target as Node) && isVisible.value) {
    isVisible.value = false;
  }
};

const toggleVisibility = () => {
  isVisible.value = !isVisible.value;
};

onMounted(() => {
  document.addEventListener('click', handleClickOutside);
});

onUnmounted(() => document.removeEventListener('click', handleClickOutside));
</script>
<template>
  <div class="dropdown" ref="dropdownElement">
    <div
      class="shared"
      :class="inElement ? 'inEl-selection' : 'selection'"
      @click="() => (isVisible = !isVisible)"
    >
      <div class="selection-title">
        <slot name="icon"></slot>
        {{ title }}
      </div>
      <div :class="{ opened: isVisible }">
        <ChevronDownVue />
      </div>
    </div>
    <div class="options" v-if="!!isVisible" @click="closeOnClick && toggleVisibility()">
      <slot name="itemsList"></slot>
    </div>
  </div>
</template>
<style scoped>
.dropdown {
  position: relative;
  width: 100%;
}

.shared {
  display: flex;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  padding: 0.625rem;
  width: 100%;
  text-align: left;
}

.selection {
  background-color: var(--neutral-0);
  border: 1px solid var(--neutral-20);
  border-radius: 0.25rem;
  box-shadow: 0.25rem 0.25rem 0.25rem var(--neutral-20);
}

.inEl-selection {
  border-bottom: 1px solid var(--neutral-20);
}

.selection-title {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.opened {
  transform: rotate(180deg);
}

.options {
  position: absolute;
  display: flex;
  flex-direction: column;
  background-color: var(--neutral-0);
  border: 1px solid var(--neutral-20);
  border-radius: 0rem;
  padding: 0.625rem;
  width: 100%;
  z-index: 1;
  box-shadow: 0.25rem 0.25rem 0.25rem var(--neutral-20);
  max-height: 25rem;
  overflow: auto;
}
</style>
