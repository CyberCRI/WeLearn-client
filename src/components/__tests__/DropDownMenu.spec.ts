import { mount } from '@vue/test-utils';
import DropDownMenu from '../DropDownMenu.vue';
import { describe, it, expect } from 'vitest';

describe('DropDownMenu', () => {
  it('renders correctly', () => {
    const wrapper = mount(DropDownMenu, {
      propsData: {
        title: 'title'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly with the items slot and click on dropdown element', async () => {
    const wrapper = mount(DropDownMenu, {
      propsData: {
        title: 'title'
      },
      slots: {
        items: '<div>Item 1</div><div>Item 2</div>'
      }
    });
    await wrapper.find('.selection').trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('.options').exists()).toBe(true);
  });
});
