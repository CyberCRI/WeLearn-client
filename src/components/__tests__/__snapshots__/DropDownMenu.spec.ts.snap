// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropDownMenu > renders correctly 1`] = `
"<div data-v-4c455bfe=\\"\\" class=\\"dropdown\\">
  <div data-v-4c455bfe=\\"\\" class=\\"shared selection\\">
    <div data-v-4c455bfe=\\"\\" class=\\"selection-title\\"> title</div>
    <div data-v-4c455bfe=\\"\\" class=\\"\\"><svg data-v-4c455bfe=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\">
        <!--!Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2024 Fonticons, Inc.-->
        <path fill=\\"currentColor\\" d=\\"M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z\\"></path>
      </svg></div>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropDownMenu > renders correctly with the items slot and click on dropdown element 1`] = `
"<div data-v-4c455bfe=\\"\\" class=\\"dropdown\\">
  <div data-v-4c455bfe=\\"\\" class=\\"shared selection\\">
    <div data-v-4c455bfe=\\"\\" class=\\"selection-title\\"> title</div>
    <div data-v-4c455bfe=\\"\\" class=\\"opened\\"><svg data-v-4c455bfe=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\">
        <!--!Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2024 Fonticons, Inc.-->
        <path fill=\\"currentColor\\" d=\\"M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z\\"></path>
      </svg></div>
  </div>
  <div data-v-4c455bfe=\\"\\" class=\\"options\\"></div>
</div>"
`;
